sql: remove restriction of foreign key on multiple outbound columns

otan opened the pull request:
Resolves https://github.com/cockroachdb/cockroach/issues/38850.

Removing the restriction of foreign keys on multiple outbound columns.
There is one minor issue we need to address which is done here -- if the
constraint name already exists, append a number next to it as postgres
does as well. We can make this fancier if we wish later.

This PR comes with a test suite of corner cases that we suspect may be
hinted at being wrong. It uses both legacy and the new optimizer
settings for foreign key checks.

Release note (sql change): We previously had a restriction that foreign
keys can only reference one outbound column at any given point in time,
e.g. in `CREATE TABLE test(a int)`, we cannot have two foreign keys
on column a. This PR removes this restriction.

otan committed:
sql: remove restriction of foreign key on multiple outbound columns

Removing the restriction of foreign keys on multiple outbound columns.
There is one minor issue we need to address which is done here -- if the
constraint name already exists, append a number next to it as postgres
does as well. We can make this fancier if we wish later.

This PR comes with a test suite of corner cases that we suspect may be
hinted at being wrong. It uses both legacy and the new optimizer
settings for foreign key checks.

Release note (sql change): We previously had a restriction that foreign
keys can only reference one outbound column at any given point in time,
e.g. in `CREATE TABLE test(a int)`, we cannot have two foreign keys
on column a. This PR removes this restriction.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/43417)
<!-- Reviewable:end -->

rohany commented:
Does this have any interaction with my pr here -- https://github.com/cockroachdb/cockroach/pull/43332/files#diff-e463e2682e19105ef8cffe2fd72b7241R975? I was worried about not being able to uniquely find fk information given an index.

otan commented:
> Does this have any interaction with my pr here -- https://github.com/cockroachdb/cockroach/pull/43332/files#diff-e463e2682e19105ef8cffe2fd72b7241R975? I was worried about not being able to uniquely find fk information given an index.

it would appear that way, and you'd need to find all matches afterwards...

rohany commented:
Is that what it means? The issue is trying to take an fk back reference on an index, that only has a table id and an index ID, and trying to find the fk ref that this corresponds to.

otan commented:
i'm kind of wondering if this is an issue already - i think this change makes it so.

by the looks of that code, it assumes one index maps to only one foreign key -- which is true if we have only 1 foreign key per outbound column. after this change, one index can now map to multiple foreign keys.

rohany commented:
That assumption only comes into play for 19.1 descriptors -- if this code comes into affect then it doesn't affect the upgrade path

thoszhang reviewed (commented):
I just realized that I don't know what is supposed to happen if we have two FK constraints on an overlapping set of columns with apparently conflicting `ON DELETE`/`ON UPDATE` behavior. What if one of them has `ON DELETE SET DEFAULT` and the other has `ON DELETE SET NULL`?

In Postgres, this happens:
```
lucy=# create table t1 (a int primary key);
CREATE TABLE
lucy=# create table t2 (a int default 1);
CREATE TABLE
lucy=# alter table t2 add constraint fk1 foreign key (a) references t1 on delete set null;
ALTER TABLE
lucy=# alter table t2 add constraint fk2 foreign key (a) references t1 on delete set default;
ALTER TABLE
lucy=# insert into t1 values (123);
INSERT 0 1
lucy=# insert into t2 values (123);
INSERT 0 1
lucy=# delete from t1 where a=123;
DELETE 1
lucy=# select * from t2;
 a
---

(1 row)
```
So it seems that `ON DELETE SET NULL` takes precedence in some sense, but I have no idea what the general principle is here. I also glanced at the Postgres documentation but couldn't find any information.

> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43417#-:-Ly-xr9TDpWVZejDnnu6:bphnxun)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/create_table.go, line 516 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43417#-Lxx8V9g7yb1hochdnbK:-Lxx8V9h7qktbVz8n3yo:b-2pljcr) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f922edf3897bf4adb68a1562256340248e06424/pkg/sql/create_table.go#L516)):*
> ```Go
> 		// If the name already exists, we try append a digit at the end and see if that exists.
> 		// We do this until we find one that is available.
> 		_, nameExists := constraintInfo[constraintName]
> ```

Should we try to pull out the logic here and in https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/alter_table.go#L349-L352 and https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/alter_table.go#L401-L404 into a separate function, now that we have 3 usages of it?
___
*[pkg/sql/create_table.go, line 519 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43417#-LxxAI-24IYp4uM17J9G:-LxxAI-24IYp4uM17J9H:bscznja) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f922edf3897bf4adb68a1562256340248e06424/pkg/sql/create_table.go#L519)):*
> ```Go
> 		nextNum := 1
> 		for nameExists {
> 			constraintName = fmt.Sprintf("fk_%s_ref_%s%d", string(d.FromCols[0]), target.Name, nextNum)
> ```

What if we started including all of the column names in the auto-generated name instead of just the first column? I think if there's one FK from `(a)` and one from `(a, b)` then `fk_a_ref_t` and `fk_a_b_ref_t` would be more informative than `fk_a_ref_t` and `fk_a_ref_t1`. (Although we'd still need the numbers, since I guess now there's nothing preventing someone from creating two indexes on the same list of columns.)
___
*[pkg/sql/create_table.go, line 637 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43417#-Ly0VK8X9rDAehH7g20Y:-Ly0VK8X9rDAehH7g20Z:b-fd17tt) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f922edf3897bf4adb68a1562256340248e06424/pkg/sql/create_table.go#L637)):*
> ```Go
> 		for _, id := range fk.OriginColumnIDs {
> 			if _, err := tbl.FindColumnByID(id); err != nil {
> 				return errors.AssertionFailedf("trying to add foreign key for column %d that doesn't exist", id)
> ```

Do we even still need this check at all?
___
*[pkg/sql/sqlbase/structured.go, line 968 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43417#-Ly-xkmv9XOeWxSspsAi:-Ly-xkmv9XOeWxSspsAj:b-ha21pq) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f922edf3897bf4adb68a1562256340248e06424/pkg/sql/sqlbase/structured.go#L968)):*
> ```Go
> 						// Found a match.
> 						// TODO(otan after PTO): write tests.
> 						desc.InboundFKs = append(
> ```

I'm not following this part. Would you mind clarifying?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> So it seems that ON DELETE SET NULL takes precedence in some sense, but I have no idea what the general principle is here. I also glanced at the Postgres documentation but couldn't find any information.


Looks like the *first* one wins - try switching the order of `set default` and `set null`.

You'll get:

```
ERROR:  insert or update on table "t2" violates foreign key constraint "fk2"
DETAIL:  Key (a)=(1) is not present in table "t1".
```

This seems to work in cockroach - but I'm curious where that is coded :)

Added a test!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43417#-:-LyAeOBO8O5JZdzTD_6p:bz00mkv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lucy-zhang and @rohany)

___
*[pkg/sql/create_table.go, line 516 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43417#-Lxx8V9g7yb1hochdnbK:-LyBUK1D-3xseiRqc6CB:b1cebpr) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f922edf3897bf4adb68a1562256340248e06424/pkg/sql/create_table.go#L516)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Should we try to pull out the logic here and in https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/alter_table.go#L349-L352 and https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/alter_table.go#L401-L404 into a separate function, now that we have 3 usages of it?
</blockquote></details>

Slightly different but I think the extra `_` in the other implementation doesn't matter!
Done!
___
*[pkg/sql/create_table.go, line 519 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43417#-LxxAI-24IYp4uM17J9G:-LyAakQ43Qt1-ohIZ6tD:bod4tfc) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f922edf3897bf4adb68a1562256340248e06424/pkg/sql/create_table.go#L519)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

What if we started including all of the column names in the auto-generated name instead of just the first column? I think if there's one FK from `(a)` and one from `(a, b)` then `fk_a_ref_t` and `fk_a_b_ref_t` would be more informative than `fk_a_ref_t` and `fk_a_ref_t1`. (Although we'd still need the numbers, since I guess now there's nothing preventing someone from creating two indexes on the same list of columns.)
</blockquote></details>

I think this is a good change that also that addresses https://github.com/cockroachdb/cockroach/issues/42208.
I think this seems like a bit of a bigger change than this PR intends - so if we want to approach this maybe in a separate PR?
___
*[pkg/sql/create_table.go, line 637 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43417#-Ly0VK8X9rDAehH7g20Y:-LyAdtU73HgWcJmV0WeT:b5fdk1s) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f922edf3897bf4adb68a1562256340248e06424/pkg/sql/create_table.go#L637)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Do we even still need this check at all?
</blockquote></details>

appears not (it's checked elsewhere!) deleting
___
*[pkg/sql/sqlbase/structured.go, line 968 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43417#-Ly-xkmv9XOeWxSspsAi:-LyAaNMx8S9syC4p1QCl:b8prsjz) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f922edf3897bf4adb68a1562256340248e06424/pkg/sql/sqlbase/structured.go#L968)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

I'm not following this part. Would you mind clarifying?
</blockquote></details>

i think we don't need this because this should only affect 19.1 -> 19.2 (as @rohany has said), and this is available 20.1 onwards
so I will just remove this.... :D


<!-- Sent from Reviewable.io -->

thoszhang reviewed (commented):
LGTM, I think it's just a matter of adding more tests now.

> Reviewed 5 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43417#-:-LyVyn8c8iI5KcZ9Q5DY:bzfezex)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/create_table.go, line 519 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43417#-LxxAI-24IYp4uM17J9G:-LyVxiKD38oDVYy0GGqV:b8uqcav) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f922edf3897bf4adb68a1562256340248e06424/pkg/sql/create_table.go#L519)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

I think this is a good change that also that addresses https://github.com/cockroachdb/cockroach/issues/42208.
I think this seems like a bit of a bigger change than this PR intends - so if we want to approach this maybe in a separate PR?
</blockquote></details>

Yeah, makes sense.
___
*[pkg/sql/create_table.go, line 639 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43417#-LyVy1vV3VpB00h3NP15:-LyVy1vV3VpB00h3NP16:byt8ny7) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4d98ba270fff441ba7fc03de2c23b898b8f3ff7/pkg/sql/create_table.go#L639)):*
> ```Go
> 	name := prefix
> 	for i := 1; nameExistsFunc(name); i++ {
> 		name = fmt.Sprintf("%s%d", prefix, i)
> ```

I might prefer `%s_%d`, to match the current (arbitrary) implementation for primary keys, but it doesn't really matter.
___
*[pkg/sql/logictest/testdata/logic_test/fk, line 2848 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43417#-LyV7-fVEeVlMT_cVkx5:-LyV7-fW1K7uPrNZHuyD:b8znumk) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4d98ba270fff441ba7fc03de2c23b898b8f3ff7/pkg/sql/logictest/testdata/logic_test/fk#L2848)):*
> ```
> DROP TABLE messages
> 
> # Also test ordering of `ON DELETE` with conflicts.
> ```

(Per offline discussion) I think we should be testing both `ON DELETE` and `ON UPDATE` with all the possible behaviors (including `NO ACTION`/`RESTRICT` and especially `CASCADE`), and also the case where e.g. there are "conflicting" FK constraints on `(a, b)` and on `(b, c)` referencing the same table with columns `(a, b, c)`, though we don't have to test every single combination in the logic tests.

Our understanding is that Postgres implements `ON DELETE/UPDATE` behavior as a series of consecutive triggers, and this is also what our current implementation does, even though its behavior for FKs from the same column hasn't really been exercised until now.


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43417)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lucy-zhang)

___
*[pkg/sql/create_table.go, line 519 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43417#-LxxAI-24IYp4uM17J9G:-LyWO8s4DEUH_-qEnNpW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f922edf3897bf4adb68a1562256340248e06424/pkg/sql/create_table.go#L519)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Yeah, makes sense.
</blockquote></details>

Done.
___
*[pkg/sql/create_table.go, line 639 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43417#-LyVy1vV3VpB00h3NP15:-LyWNx21-IOPQPGJ_45k:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4d98ba270fff441ba7fc03de2c23b898b8f3ff7/pkg/sql/create_table.go#L639)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

I might prefer `%s_%d`, to match the current (arbitrary) implementation for primary keys, but it doesn't really matter.
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/fk, line 2848 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43417#-LyV7-fVEeVlMT_cVkx5:-LyWNzIf6QuU5RYJlcV_:b-nqlq6g) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4d98ba270fff441ba7fc03de2c23b898b8f3ff7/pkg/sql/logictest/testdata/logic_test/fk#L2848)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

(Per offline discussion) I think we should be testing both `ON DELETE` and `ON UPDATE` with all the possible behaviors (including `NO ACTION`/`RESTRICT` and especially `CASCADE`), and also the case where e.g. there are "conflicting" FK constraints on `(a, b)` and on `(b, c)` referencing the same table with columns `(a, b, c)`, though we don't have to test every single combination in the logic tests.

Our understanding is that Postgres implements `ON DELETE/UPDATE` behavior as a series of consecutive triggers, and this is also what our current implementation does, even though its behavior for FKs from the same column hasn't really been exercised until now.
</blockquote></details>

Cockroach actually does SET DEFAULT/NULL/CASCADE in order, and evaluates RESTRICT/NO ACTION after all row updates have occurred at the end of the statement. Tests reflect this.


<!-- Sent from Reviewable.io -->

thoszhang reviewed (commented):
> Reviewed 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43417)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/logictest/testdata/logic_test/fk, line 2848 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43417#-LyV7-fVEeVlMT_cVkx5:-LyZxP4KCaBW1G1G6GDv:bb2hviw) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4d98ba270fff441ba7fc03de2c23b898b8f3ff7/pkg/sql/logictest/testdata/logic_test/fk#L2848)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Cockroach actually does SET DEFAULT/NULL/CASCADE in order, and evaluates RESTRICT/NO ACTION after all row updates have occurred at the end of the statement. Tests reflect this.
</blockquote></details>

Hmm, what I originally vaguely had in mind was that we'd be able to write an end-to-end test that essentially does what the script does, as opposed to putting the output in the logic tests themselves. But I don't even know if that's possible.

Anyway, I am somewhat conflicted about adding ~900 lines of exhaustive, programmatically generated logic tests. On the one hand, there is some precedent for covering "every case" of something (e.g., earlier in the `fk` file we have every combination of inserts involving NULLs and composite FKs with different matching behavior, which is why the file is already almost 3000 lines long), and there is some sense in which every case tested is genuinely different from the others.

On the other hand, I don't like the idea of either the script and the logic test getting out of sync _or_ having to update the script to update the logic tests. Also, the fk test file is already one of the longest-running logic test files, so I'm reluctant to make it even longer.

I think the logic tests are more useful for hand-generated edge cases and regression tests than for exhaustively covering a space of possibilities. (Arguably we've already violated this with the current logic tests.) I think we probably agree that some kind of randomized testing where we can verify results would be ideal, but given that we don't have that....I'm not sure whether this is a good stopgap measure.

Is there any way to pare this down to a more manageable set of cases that's more human-readable/editable? For instance, I think some of the UPDATE vs. DELETE cases are in some sense equivalent, and SET NULL vs. SET DEFAULT are also similar in some cases.


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/sql/logictest/testdata/logic_test/fk, line 2848 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43417#-LyV7-fVEeVlMT_cVkx5:-Ly_3jlBES9aXAmk67iq:b-twr0mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4d98ba270fff441ba7fc03de2c23b898b8f3ff7/pkg/sql/logictest/testdata/logic_test/fk#L2848)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Hmm, what I originally vaguely had in mind was that we'd be able to write an end-to-end test that essentially does what the script does, as opposed to putting the output in the logic tests themselves. But I don't even know if that's possible.

Anyway, I am somewhat conflicted about adding ~900 lines of exhaustive, programmatically generated logic tests. On the one hand, there is some precedent for covering "every case" of something (e.g., earlier in the `fk` file we have every combination of inserts involving NULLs and composite FKs with different matching behavior, which is why the file is already almost 3000 lines long), and there is some sense in which every case tested is genuinely different from the others.

On the other hand, I don't like the idea of either the script and the logic test getting out of sync _or_ having to update the script to update the logic tests. Also, the fk test file is already one of the longest-running logic test files, so I'm reluctant to make it even longer.

I think the logic tests are more useful for hand-generated edge cases and regression tests than for exhaustively covering a space of possibilities. (Arguably we've already violated this with the current logic tests.) I think we probably agree that some kind of randomized testing where we can verify results would be ideal, but given that we don't have that....I'm not sure whether this is a good stopgap measure.

Is there any way to pare this down to a more manageable set of cases that's more human-readable/editable? For instance, I think some of the UPDATE vs. DELETE cases are in some sense equivalent, and SET NULL vs. SET DEFAULT are also similar in some cases.
</blockquote></details>

I couldn't really find a nice way of doing a end2end test way - if there is would be happy to change it. Perhaps worth investing in a framework of programatic logic tests...

I've simplified to ten permutations of DELETE and UPDATE, and updated it accordingly :)


<!-- Sent from Reviewable.io -->

thoszhang reviewed (approved):
> Reviewed 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43417)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/logictest/testdata/logic_test/fk, line 2848 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43417#-LyV7-fVEeVlMT_cVkx5:-Ly_DKs5C5h_BkPkNrG7:bb73djm) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4d98ba270fff441ba7fc03de2c23b898b8f3ff7/pkg/sql/logictest/testdata/logic_test/fk#L2848)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

I couldn't really find a nice way of doing a end2end test way - if there is would be happy to change it. Perhaps worth investing in a framework of programatic logic tests...

I've simplified to ten permutations of DELETE and UPDATE, and updated it accordingly :)
</blockquote></details>

SGTM
___
*[pkg/sql/logictest/testdata/logic_test/fk, line 2848 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/43417#-Ly_DJXN6AHiIbMZQfJv:-Ly_DJXN6AHiIbMZQfJw:b-gb4ae3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ad7a9f80d7d44d0885b001d0b7870d1f269f172/pkg/sql/logictest/testdata/logic_test/fk#L2848)):*
> ```
> DROP TABLE messages
> 
> # Test conflicting foreign keys ON DELETE and ON UPDATE - some known corners cases.
> ```

nit: `corner`
___
*[pkg/sql/logictest/testdata/logic_test/fk, line 2849 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/43417#-Ly_DQ5W4qXC1YP6dVdw:-Ly_DQ5W4qXC1YP6dVdx:b-rh7jpp) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ad7a9f80d7d44d0885b001d0b7870d1f269f172/pkg/sql/logictest/testdata/logic_test/fk#L2849)):*
> ```
> 
> # Test conflicting foreign keys ON DELETE and ON UPDATE - some known corners cases.
> # Tests were generated from https://github.com/otan-cockroach/otan-scripts/blob/master/38850.
> ```

I think if we leave the URL in here we may also want to mention that the tests aren't expected to be updated using the script.


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43417)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lucy-zhang)

___
*[pkg/sql/logictest/testdata/logic_test/fk, line 2848 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43417#-LyV7-fVEeVlMT_cVkx5:-Ly_EZFe-E7V-UzsfZfq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4d98ba270fff441ba7fc03de2c23b898b8f3ff7/pkg/sql/logictest/testdata/logic_test/fk#L2848)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

SGTM
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/fk, line 2848 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/43417#-Ly_DJXN6AHiIbMZQfJv:-Ly_EZdI5AX6zOeeRZpk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ad7a9f80d7d44d0885b001d0b7870d1f269f172/pkg/sql/logictest/testdata/logic_test/fk#L2848)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

nit: `corner`
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/fk, line 2849 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/43417#-Ly_DQ5W4qXC1YP6dVdw:-Ly_EaDJ2vQuEJjdUstP:b-ie3qie) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ad7a9f80d7d44d0885b001d0b7870d1f269f172/pkg/sql/logictest/testdata/logic_test/fk#L2849)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

I think if we leave the URL in here we may also want to mention that the tests aren't expected to be updated using the script.
</blockquote></details>

I'll remove it, as it's no longer really needed to be known.


<!-- Sent from Reviewable.io -->

otan commented:
bors r=lucy-zhang

tftr
